nemo/collections/asr/parts/utils/online_clustering.py

```diff
@@ -0,0 +1,1106 @@
+
+        if (org_size - merge_quantity) != merged_embs.shape[0]:
+            raise ValueError(
+                f"Reducer output {merged_embs.shape[0]} is not matched to the target quantity {org_size - merge_quantity}."
+            )
+
+    else:
+        merged_embs = pre_embs[target_emb_index]
+        merged_clus_labels = pre_clus_labels[target_emb_index]
+        index_mapping = (target_emb_index, torch.arange(0))
+    return merged_embs, merged_clus_labels, index_mapping
+
+
+@torch.jit.script
+def get_first_arg_index(mat: torch.Tensor, label: int) -> int:
+    """
+    Get the index of the first element are specified by `index` variable.
+
+    Args:
+        mat (Tensor):
+            Source matrix filled with indices
+        label (int):
+            Label which we want to find the first occuring index
+
+    Returns:
+        (int) The first index of the given label
+    """
+    return int(torch.where(mat == label)[0][0])
+
+
+class OnlineSpeakerClustering:
+    """
+    Online clustering method for speaker diarization based on cosine similarity.
+
+    Regular Clustering Attributes:
+
+        max_num_speakers (int):
+            The upper bound for the number of speakers in each session
+        max_rp_threshold (float):
+            Limits the range of parameter search.
+            Clustering performance can vary depending on this range.
+            Default is 0.15.
+        enhanced_count_thres (int):
+            For the short audio recordings, clustering algorithm cannot
+            accumulate enough amount of speaker profile for each cluster.
+            Thus, function `getEnhancedSpeakerCount` employs anchor embeddings
+            (dummy representations) to mitigate the effect of cluster sparsity.
+            enhanced_count_thres = 40 is recommended.
+        sparse_search_volume (int):
+            Number of p_values we search during NME analysis.
+            Default is 30. The lower the value, the faster NME-analysis becomes.
+            Lower than 20 might cause a poor parameter estimation.
+        fixed_thres (float):
+            If fixed_thres value is provided, NME-analysis process will be skipped.
+            This value should be optimized on a development set to obtain a quality result.
+            Default is None and performs NME-analysis to estimate the threshold.
+        min_samples_for_nmesc (int):
+            The minimum number of samples required for NME clustering. This avoids
+            zero p_neighbour_lists. If the input has fewer segments than min_samples,
+            it is directed to the enhanced speaker counting mode.
+        sparse_search (bool):
+            Toggle sparse search mode. If True, limit the size of p_value_list to sparse_search_volume.
+        cuda (bool):
+            Use cuda for Eigen decomposition if cuda=True.
+        device (torch.device):
+            Torch device variable
+
+    Online Processing Attributes:
+
+        history_buffer_size (int):
```

Comment:
Did the default value of 150 come based on experiments results of accuracy or performance?

Reply:
This value is tested on AMI dev. Will update later if it is needed.